services/placement/placement.go

```diff
@@ -112,9 +115,10 @@ func (p *placement) SetInstances(instances []services.PlacementInstance) service
 		sort.Sort(ByIDAscending(instances))
 	}
 
-	p.instancesByShard = instancesByShard
-	p.instances = instancesMap
-	return p
+	newp := *p
```

Comment:
Hm why the change of semantics? If we do decide to do this, we need to find all the places where these APIs are used and make sure we replace `p.SetXXX(...)` with `p = p.SetXXX(...)`. Sounds like a lot of work so prefer we don't do it unless we have a good reason to do so.